build - Fix compile errors in 1.6 stream that occurs with g++-11

OlofGullnas opened the pull request:
Compiling with g++-11 produces errors (or warnings treated as errors in the build) that did not occur with gcc-9.

The compiler I used showed "gcc version 11.3.0 (SUSE Linux)" when invoked with "-v".

Three problems fixed:
1) Missing #include <optional> in common/communication/socket.h.
   There was no type std::optional. Compiler suggested adding the
   include. (Header probably indirectly included with earlier versions.)

2) Warning about possible use of uninitialized variable when compiling
   service/source/manager/handle.cpp near line 795. This warning only
   showed up in builds without --debug! Introducing a local variable
   to hold the result of a call to
      algoritm::sorted::intersection()
   made the warning go away. I have not investigated if the warning
   is correct or a compiler bug (unlikely:-).
   The warning did not occur in g++-9. With g++-10 the warning occured
   but with a message that provided essentially no context.
   In g++-11 the message was much improved.

3) fix waring about integer compare with different signedness.
   (This warning has also appeared for some users with g++-9
   after the introduction of container::sorted::Set)

   test_set.cpp had three instances of integer compares with different
   signedness. Something like:
      set.size() == origin.size()
   where "set" is a container::sorted::Set and "origin" is constructed with
   array::make().

   Replaced with
     range::size( set) == range::size( origin)

OlofGullnas committed:
build - Fix compile errors in 1.6 stream that occurs with g++-11

Compiling with g++-11 produces errors (or warnings treated as errors
in the build) that did not occur with gcc-9.

The compiler I used showed "gcc version 11.3.0 (SUSE Linux)" when
invoked with "-v".

Three problems fixed:
1) Missing #include <optional> in common/communication/socket.h.
   There was no type std::optional. Compiler suggested adding the
   include. (Header probably indirectly included with earlier versions.)

2) Warning about possible use of uninitialized variable when compiling
   service/source/manager/handle.cpp near line 795. This warning only
   showed up in builds without --debug! Introducing a local variable
   to hold the result of a call to
      algoritm::sorted::intersection()
   made the warning go away. I have not investigated if the warning
   is correct or a compiler bug (unlikely:-).
   The warning did not occur in g++-9. With g++-10 the warning occured
   but with a message that provided essentially no context.
   In g++-11 the message was much improved.

3) fix waring about integer compare with different signedness.
   (This warning has also appeared for some users with g++-9
   after the introduction of container::sorted::Set)

   test_set.cpp had three instances of integer compares with different
   signedness. Something like:
      set.size() == origin.size()
   where "set" is a container::sorted::Set and "origin" is constructed with
   array::make().

   Replaced with
     range::size( set) == range::size( origin)

lazan commented on middleware/service/source/manager/handle.cpp:
```diff
@@ -791,7 +791,20 @@ namespace casual
+                        // involved code. 
+                        //for( auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
+                        auto intersect = algorithm::sorted::intersection( message.content.services(), reply.content.services());
+                        for( auto& name : std::get< 1>( intersect))
```
I would like you to try  `for( const auto& name : std::get< 1>( <same-as-the-original>)`

const ref, should extend the lifetime of  name in "whatever the compiler generates for the range-for-loop".

OlofGullnas replied:
@lazan 
Tried with following code:
```
                        for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
                           if( auto found = algorithm::find( state.reverse_routes, name))
                              reply.routes.services.emplace_back( found->first, found->second);

                        local::optional::send( state, message.process.ipc, reply);


```
Same error:

compile: /export/home/guo/casual/middleware/service/source/manager/handle.cpp
error:
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::begin() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:94:61: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_first.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   94 |          constexpr iterator begin() const noexcept { return m_first;}
      |                                                             ^~~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::end() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:95:59: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_last.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   95 |          constexpr iterator end() const noexcept { return m_last;}
      |                                                           ^~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

aborting due to errors
error building...
[

OlofGullnas replied:
Very long error message. Not so easy to see where/why the warning(error) occurs. Line 805 in the lambda mentioned at the end of the messages is the line "for( const aiuo& name: std::get< 1>(...", but the warnings occur in  casual::common::Range::begin() and  casual::common::Range::end(). 
The "real" problem may be "deeper down" than the for loop. If there IS a problem and not a compiler bug :-). Compiler bugs are not common, and as the error occurs with both gcc-10 and gcc-11 so the bug has been there for some time if there really is a problem. Makes a compiler bug less likely.
Guessing from the message there is some agressive inlining going on in the compiler...

OlofGullnas replied:
on the  "especially for debug-mode."...
This particular warning does NOT occur when compiling with --debug. It only occurs when --debug is NOT used. Likely because the compiler is not doing the same level of data flow analysis when compiling with debug. I noticed that compiling this file takes significant time, around 10 seconds or so, so the compiler does a significant amount of work :-)

The warning about comparing integer expressions with different "signage" in test_set.cpp only occurs with --debug when using g++-9:

- g++-9 without --debug: no warning
- g++-9 with --debug: warning
- g++-11 with or without --debug: warning
That building with debug makes a difference is a bit surprising :-) Could understand it if the comparison was in an assert that is conditional on NDEBUG. But in this case the "compare" should be "unconditional". Unless the GTEST framework is doing something really odd...

lazan replied:
Ahh, I got it backwards :)

lazan replied:
Can you paste the error again within  _three ticks_ (markdown syntax to make it "termialish")

OlofGullnas replied:
@lazan 
```compile: /export/home/guo/casual/middleware/service/source/manager/handle.cpp
error:
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::begin() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:94:61: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_first.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   94 |          constexpr iterator begin() const noexcept { return m_first;}
      |                                                             ^~~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../common/include/common/view/binary.h:10,
                 from ../common/include/common/serialize/traits.h:15,
                 from ../common/include/common/serialize/macro.h:10,
                 from ../common/include/common/strong/type.h:11,
                 from ../common/include/common/message/type.h:12,
                 from ../common/include/common/message/service.h:10,
                 from include/service/manager/state.h:9,
                 from include/service/manager/handle.h:12,
                 from /export/home/guo/casual/middleware/service/source/manager/handle.cpp:8:
In member function ‘constexpr casual::common::Range<Iter>::iterator casual::common::Range<Iter>::end() const [with Iter = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’,
    inlined from ‘casual::service::manager::handle::local::{anonymous}::domain::discovery::internal::request(casual::service::manager::State&)::<lambda(casual::domain::message::discovery::internal::Request&)>’ at /export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:149:
../common/include/common/range.h:95:59: error: ‘<unnamed>.std::tuple<casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<0, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Tuple_impl<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >::<unnamed>.std::_Head_base<1, casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >, false>::_M_head_impl.casual::common::Range<__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > > >::m_last.__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >::_M_current’ may be used uninitialized [-Werror=maybe-uninitialized]
   95 |          constexpr iterator end() const noexcept { return m_last;}
      |                                                           ^~~~~~
/export/home/guo/casual/middleware/service/source/manager/handle.cpp: In lambda function:
/export/home/guo/casual/middleware/service/source/manager/handle.cpp:805:94: note: ‘<anonymous>’ declared here
  805 |                         for( const auto& name : std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services())))
      |                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

aborting due to errors
error building...
[infra] guo@bvin01-unisb-01:~/casual $```

OlofGullnas replied:
That do make the message look better ;-)

OlofGullnas replied:
@lazan 
I saw the following in a list of new C++ 20 features [https://github.com/AnthonyCalandra/modern-cpp-features#range-based-for-loop-with-initializer](https://github.com/AnthonyCalandra/modern-cpp-features#range-based-for-loop-with-initializer) . Is this related to the warning I get with g++-11? I.e the "...elegant solution to a common lifetime problem". Introducing a local variable before the for is somewhat related in effect to what I assume this new language featuire does. 

Range-based for loop with initializer
This feature simplifies common code patterns, helps keep scopes tight, and offers an elegant solution to a common lifetime problem.

for (auto v = std::vector{1, 2, 3}; auto& e : v) {
  std::cout << e;
}
// prints "123"

OlofGullnas replied:
I have been looking at [https://en.cppreference.com/w/cpp/language/range-for](https://en.cppreference.com/w/cpp/language/range-for) for a while. In particular the section about "temporary range expression". 

"If range-expression returns a temporary, its lifetime is extended until the end of the loop, as indicated by binding to the forwarding reference __range. Lifetimes of all temporaries within range-expression are _not (until C++23)_ extended if they would otherwise be destroyed at the end of range-expression (since C++23)."

This problem may be worked around using init-statement:

for (T thing = foo(); auto& x : thing.items()) { /* ... */ } // OK (since C++20)"


The init statement solution is semantically very like the effect of the local variable I added. 
 
I am to tired to think through everything now (and I am not a real C++ specialist), but I get the feeling that lifetimes are at play here. Intersection() returns a tuple that is a temporary, and std::get< 1> also returns a temporary (does it?) based on item 1 in the tuple. Now it is possible that the code is "correct" but I get the feeling that the compiler for some reason or another thinks otherwise.

It does say "MAY be used uninitialized", If taken literally (!) this should mean that there exist a path where it may be used uninitialized, but it is not always uninitialized. 

This is the single place in all of casual that gets this warning, so something is special/unique with this code. 
Should probably make up small examples and try them out.

lazan reviewed (changes requested):
I would like you to try my suggestion. I'm not that keen on introducing _variables_ to get around "possible compiler bugs", especially for _debug-mode_. 

Another thing, unless we're actually building on a givet compiler, these things will happen, I think we have to decide what compiler we "support", and thats it.  And only fix stuff that we know is not conformant (the diffrent size types, for example).

lazan commented:
@OlofGullnas Something like this should work. And one can argue that it's easier to read:

```cpp 
   // 'lookup' routes for services that caller are interested in.
   {
        const auto difference = std::get< 1>( algorithm::sorted::intersection( message.content.services(), reply.content.services()));

         // check if some of the rest has _routes_
         for( auto& name : difference)
            if( auto found = algorithm::find( state.reverse_routes, name))
               reply.routes.services.emplace_back( found->first, found->second);
   }

   local::optional::send( state, message.process.ipc, reply);

```
                        
I don't believe we need comments about compiler errors in the code. Weill maybe sometime :) but not for this.

What do you think?

OlofGullnas committed:
Service Manager - Simplify code that caused a "may be used uninitialized" warning.

New code is simpler, clearer and easier to understand! Also avoids the warning.

OlofGullnas commented:
Implemented your suggested change, i.e. to introduced a "differences" local variable that holds the range to iterate over...

The code is certainly easier to read than the original code. And it does not cause the warning. Avoids any issues with lifetimes that may cause problems or confuse the compiler

Personally I am not very fond of very long lines. I like to format them as multiple lines, or split into multiple statements. I also think that introducing local variables with good names  to hold intermediate results can split complex (long)  operations into smallert steps and provide more clues/information about what is being done.

lazan reviewed (approved):
Nice. I think this works. Good work.
